compiler/luci/pass/src/RequantizePass.cpp

```diff
@@ -0,0 +1,276 @@
+// Original: -127 ~ 127 zp = 0
+// After requantization: 1 ~ 255 zp = 128
+void wrequant_int8_to_uint8(CircleConst *node)
+{
+  assert(node->dtype() == loco::DataType::S8);
+
+  uint32_t size = node->size<loco::DataType::S8>();
+  std::vector<int32_t> requantized_values(size);
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    int32_t data = node->at<loco::DataType::S8>(i);
+    requantized_values[i] = data + 128;
+  }
+
+  node->dtype(loco::DataType::U8); // change the type of tensor
+  node->size<loco::DataType::U8>(size);
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    assert(0 <= requantized_values[i] && requantized_values[i] <= 255);
+    node->at<loco::DataType::U8>(i) = requantized_values[i];
+  }
+}
+
+// Check if node is weights of conv2d, depthwise_conv2d, transpose_conv, or fully_connected layer
+bool is_weights(CircleNode *node)
+{
+  auto circle_const = dynamic_cast<CircleConst *>(node);
+  if (circle_const == nullptr)
+    return false;
+
+  auto succs = loco::succs(node);
+  if (succs.size() != 1) // assume weights is used by only one node
+    return false;
+
+  for (auto out : succs)
+  {
+    auto conv = dynamic_cast<CircleConv2D *>(out);
+    if (conv != nullptr && conv->filter() == circle_const)
+      return true;
+
+    auto dw_conv = dynamic_cast<CircleDepthwiseConv2D *>(out);
+    if (dw_conv != nullptr && dw_conv->filter() == circle_const)
+      return true;
+
+    auto t_conv = dynamic_cast<CircleTransposeConv *>(out);
+    if (t_conv != nullptr && t_conv->filter() == circle_const && circle_const->rank() == 4)
+      return true;
+
+    auto fc = dynamic_cast<CircleFullyConnected *>(out);
+    if (fc != nullptr && fc->weights() == circle_const)
+      return true;
+  }
+  return false;
+}
+
+/**
+ * @brief RequantizeActivation requantizes tensors for activations
+ */
+struct RequantizeActivation final : public luci::CircleNodeMutableVisitor<bool>
+{
+  RequantizeActivation(loco::DataType input, loco::DataType output)
+      : _input_type(input), _output_type(output)
+  {
+  }
+
+  loco::DataType _input_type;
+  loco::DataType _output_type;
+
+  // Requantize input tensors of each node
+  bool visit(luci::CircleNode *node)
+  {
+    LOGGER(l);
+    INFO(l) << "RequantizeActivation visit node: " << node->name() << std::endl;
+    auto arity = node->arity();
+    for (uint32_t i = 0; i < arity; i++)
+    {
+      auto input_node = node->arg(i);
+      auto circle_node = loco::must_cast<luci::CircleNode *>(input_node);
+
+      // Check if this was already requantized
+      if (circle_node->dtype() == _output_type)
+        continue;
+
+      // Check if this was not quantized (only quantized tensors are requantized)
+      if (circle_node->quantparam() == nullptr)
+        continue;
+
+      if (_input_type == loco::DataType::S8 && _output_type == loco::DataType::U8)
+      {
+        // Check if this is not bias (bias is not requantized when int8 -> uint8)
+        auto iw = get_input_weight_of_bias(circle_node);
```

Comment:
Could you explain about `get_input_weight_of_bias` function more detail? It seems there is no other place that uses this function. But then, return value doesn't seem necessary. I mean, just true, false values are enough.

Reply:
You're right. The return values are not used in this pass. They are used to calculate the scales of bias in QuantizedWithMinMaxPass. I just copied the function from the source code. I've updated the code to remove the redundant codes.

Reply:
Also, I've updated the code to correctly handle const tensors (not weights nor bias). PTAL.